File: packages/go/capture_test.go
package crumbtrail

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"sync"
	"testing"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

type memorySink struct {
	mu      sync.Mutex
	batches []Batch
}

func (s *memorySink) Enqueue(b Batch) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.batches = append(s.batches, b)
	return true
}
func (s *memorySink) events() []Event {
	s.mu.Lock()
	defer s.mu.Unlock()
	var out []Event
	for _, b := range s.batches {
		out = append(out, b.Events...)
	}
	return out
}
func request(body string) *http.Request {
	r := httptest.NewRequest("POST", "https://example.com/api/quote?token=secret", strings.NewReader(body))
	r.Header.Set("Content-Type", "application/json")
	r.Header.Set("x-crumbtrail-session-id", "go-contract-session")
	r.Header.Set("x-crumbtrail-request-id", "go-contract-request")
	return r
}
func find(t *testing.T, s *memorySink, kind string) map[string]any {
	t.Helper()
	for _, e := range s.events() {
		if e.K == kind {
			return e.D
		}
	}
	t.Fatalf("missing %s", kind)
	return nil
}
func TestLoopbackHTTPDatabase(t *testing.T) {
	db, err := sql.Open("sqlite3", ":memory:")
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	db.SetMaxOpenConns(1)
	observed, err := WrapDB(db, "sqlite")
	if err != nil {
		t.Fatal(err)
	}
	sink := &memorySink{}
	handler := Middleware(Options{sink, "go-test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		input, err := io.ReadAll(r.Body)
		if err != nil {
			t.Error(err)
		}
		if !strings.Contains(string(input), "never-send") {
			t.Error("request changed")
		}
		tx, err := observed.BeginTx(r.Context(), nil)
		if err != nil {
			t.Error(err)
			return
		}
		if _, err = tx.ExecContext(r.Context(), "CREATE TABLE items (value INTEGER)"); err != nil {
			t.Error(err)
		}
		stmt, err := tx.PrepareContext(r.Context(), "INSERT INTO items(value) VALUES (?)")
		if err != nil {
			t.Error(err)
			return
		}
		if _, err = stmt.ExecContext(r.Context(), 731); err != nil {
			t.Error(err)
		}
		stmt.Close()
		if err = tx.Commit(); err != nil {
			t.Error(err)
		}
		var operand int
		if err = observed.QueryRowContext(r.Context(), "SELECT value FROM items").Scan(&operand); err != nil || operand != 731 {
			t.Errorf("database result %v %d", err, operand)
		}
		if _, err = observed.ExecContext(r.Context(), "SELECT * FROM absent_secret_table"); err == nil {
			t.Error("expected error")
		}
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{"total":37.5,"currency":"CAD","password":"never-send"}`)
	}))
	server := httptest.NewServer(handler)
	defer server.Close()
	r := request(`{"amount":18.75,"entityId":731,"currency":"CAD","password":"never-send"}`)
	r.URL.Scheme = "http"
	r.URL.Host = strings.TrimPrefix(server.URL, "http://")
	r.RequestURI = ""
	response, err := server.Client().Do(r)
	if err != nil {
		t.Fatal(err)
	}
	body, _ := io.ReadAll(response.Body)
	response.Body.Close()
	if !strings.Contains(string(body), "never-send") {
		t.Fatal("response changed")
	}
	start := find(t, sink, "backend.req.start")
	end := find(t, sink, "backend.req.end")
	if start["requestBodyState"] != "redacted" || end["responseBodyState"] != "redacted" {
		t.Fatal(start, end)
	}
	if !strings.Contains(start["body"].(string), "18.75") || !strings.Contains(end["responseBody"].(string), "37.5") {
		t.Fatal("safe operands lost")
	}
	find(t, sink, "db.statement")
	find(t, sink, "db.error")
	encoded, _ := json.Marshal(sink.batches)
	if strings.Contains(string(encoded), "never-send") || strings.Contains(string(encoded), "absent_secret_table") {
		t.Fatal("private evidence leaked")
	}
	if output := os.Getenv("CRUMBTRAIL_CAPTURE_CONTRACT_OUTPUT"); output != "" {
		if err := os.WriteFile(output, contractJSON(sink), 0600); err != nil {
			t.Fatal(err)
		}
	}
}
func TestBodyPolicy(t *testing.T) {
	for _, raw := range []string{`{"a":1,"a":2}`, `{"a":`, strings.Repeat("[", 10) + strings.Repeat("]", 10)} {
		if got := captureBody([]byte(raw), false); got.State != "invalid" {
			t.Fatal(raw, got)
		}
	}
	got := captureBody([]byte(`{"amount":18.75,"email":"x@example.com","routing_number":123,"id":4111111111111111}`), false)
	if got.State != "redacted" || strings.Contains(got.Body.(string), "4111111111111111") {
		t.Fatal(got)
	}
	if got := captureBody(make([]byte, bodyLimit+1), false); got.State != "truncated" {
		t.Fatal(got)
	}
}
func TestTruncationInterfacesAndFailureIsolation(t *testing.T) {
	sink := &memorySink{}
	large := `{"data":"` + strings.Repeat("a", 20000) + `"}`
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if _, ok := w.(http.Flusher); !ok {
			t.Error("flusher lost")
		}
		b, _ := io.ReadAll(r.Body)
		if string(b) != large {
			t.Error("request changed")
		}
		w.Header().Set("Content-Type", "application/json")
		io.Copy(w, strings.NewReader(large))
	}))
	recorder := httptest.NewRecorder()
	handler.ServeHTTP(recorder, request(large))
	if recorder.Body.String() != large {
		t.Fatal("response changed")
	}
	if find(t, sink, "backend.req.start")["requestBodyState"] != "truncated" || find(t, sink, "backend.req.end")["responseBodyState"] != "truncated" {
		t.Fatal("missing truncation")
	}
	bad := Middleware(Options{panicSink{}, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { io.WriteString(w, "ok") }))
	bad.ServeHTTP(httptest.NewRecorder(), request(""))
}

type panicSink struct{}

func (panicSink) Enqueue(Batch) bool { panic("sink failed") }
func TestPanicPreserved(t *testing.T) {
	sink := &memorySink{}
	original := errors.New("private")
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(http.ResponseWriter, *http.Request) { panic(original) }))
	func() {
		defer func() {
			if recover() != original {
				t.Error("panic changed")
			}
		}()
		handler.ServeHTTP(httptest.NewRecorder(), request(""))
	}()
	find(t, sink, "backend.req.error")
}
func TestDefaultAndInvalidCorrelation(t *testing.T) {
	sink := &memorySink{}
	for _, options := range []Options{{Sink: sink}, {Sink: sink, ShouldCapture: func(*http.Request) bool { return true }}} {
		r := request("")
		r.Header.Set("x-crumbtrail-request-id", "invalid\n")
		Middleware(options)(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(204) })).ServeHTTP(httptest.NewRecorder(), r)
	}
	if len(sink.events()) != 0 {
		t.Fatal("unexpected capture")
	}
}
func TestSenderTLSRetryAndClose(t *testing.T) {
	var mu sync.Mutex
	count := 0
	server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		count++
		n := count
		mu.Unlock()
		if r.URL.Path != "/api/events" || r.Header.Get("Authorization") != "Bearer test-key" {
			t.Error("invalid delivery")
		}
		var batch Batch
		if err := json.NewDecoder(r.Body).Decode(&batch); err != nil {
			t.Error(err)
		}
		if n == 1 {
			w.WriteHeader(404)
		} else {
			w.WriteHeader(200)
		}
	}))
	defer server.Close()
	sender, err := NewSender(SenderConfig{server.URL, "test-key", server.Client()})
	if err != nil {
		t.Fatal(err)
	}
	if !sender.Enqueue(Batch{"session", []Event{{1, "backend.req.end", map[string]any{"requestId": "request"}}}}) {
		t.Fatal("enqueue failed")
	}
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	if err := sender.Close(ctx); err != nil {
		t.Fatal(err)
	}
	mu.Lock()
	defer mu.Unlock()
	if count != 2 {
		t.Fatal(count)
	}
	if sender.Enqueue(Batch{}) {
		t.Fatal("enqueue after close")
	}
}
func TestSenderRejectsEndpoints(t *testing.T) {
	for _, endpoint := range []string{"http://localhost", "https://user:pass@example.com", "https://example.com/?key=x"} {
		if _, err := NewSender(SenderConfig{Endpoint: endpoint, Key: "key"}); err == nil {
			t.Fatal(endpoint)
		}
	}
}

func contractJSON(s *memorySink) []byte {
	b, _ := json.Marshal(Batch{s.batches[0].SessionID, s.events()})
	return b
}

func TestPanicAfterWriteKeepsStatus(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(202)
		io.WriteString(w, "partial")
		panic("private failure")
	}))
	func() { defer func() { recover() }(); handler.ServeHTTP(httptest.NewRecorder(), request("")) }()
	if find(t, sink, "backend.req.end")["statusCode"] != 202 {
		t.Fatal("recorded status differs from sent status")
	}
	find(t, sink, "backend.req.error")
}
func TestConcurrentContextsStaySeparate(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { time.Sleep(time.Millisecond); io.WriteString(w, "ok") }))
	var group sync.WaitGroup
	for _, id := range []string{"request-one", "request-two"} {
		group.Add(1)
		go func(id string) {
			defer group.Done()
			r := request("")
			r.Header.Set("x-crumbtrail-request-id", id)
			handler.ServeHTTP(httptest.NewRecorder(), r)
		}(id)
	}
	group.Wait()
	counts := map[string]int{}
	for _, e := range sink.events() {
		counts[e.D["requestId"].(string)]++
	}
	if counts["request-one"] != 2 || counts["request-two"] != 2 {
		t.Fatal(counts)
	}
}
func TestSenderDoesNotFollowRedirect(t *testing.T) {
	redirected := false
	destination := httptest.NewTLSServer(http.HandlerFunc(func(http.ResponseWriter, *http.Request) { redirected = true }))
	defer destination.Close()
	server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { http.Redirect(w, r, destination.URL, 307) }))
	defer server.Close()
	sender, err := NewSender(SenderConfig{server.URL, "key", server.Client()})
	if err != nil {
		t.Fatal(err)
	}
	sender.Enqueue(Batch{})
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	if err := sender.Close(ctx); err != nil {
		t.Fatal(err)
	}
	if redirected {
		t.Fatal("redirect leaked key")
	}
}

func TestEventLimitPreservesBoundaries(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		for i := 0; i < 250; i++ {
			recordSQL(r.Context(), "sqlite", "SELECT 1", time.Now(), nil, nil)
		}
		io.WriteString(w, "ok")
	}))
	handler.ServeHTTP(httptest.NewRecorder(), request(""))
	find(t, sink, "backend.req.start")
	find(t, sink, "backend.req.end")
	gap := find(t, sink, "capture_gap")
	if gap["droppedEvents"] != 50 {
		t.Fatal(gap)
	}
	if len(sink.events()) != 203 {
		t.Fatal(len(sink.events()))
	}
}

type fromWriter struct{ *httptest.ResponseRecorder }

func (w fromWriter) ReadFrom(r io.Reader) (int64, error) { return io.Copy(w.ResponseRecorder, r) }

type readerOnly struct{ io.Reader }

func TestReaderFromPreservedAndCaptured(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if _, ok := w.(io.ReaderFrom); !ok {
			t.Error("ReaderFrom lost")
		}
		w.Header().Set("Content-Type", "application/json")
		io.Copy(w, readerOnly{strings.NewReader(`{"total":37.5}`)})
	}))
	recorder := httptest.NewRecorder()
	handler.ServeHTTP(fromWriter{recorder}, request(""))
	if recorder.Body.String() != `{"total":37.5}` {
		t.Fatal(recorder.Body.String())
	}
	if find(t, sink, "backend.req.end")["responseBodyState"] != "captured" {
		t.Fatal("ReaderFrom bypassed capture")
	}
}

func TestPartialRequestDoesNotInventOperand(t *testing.T) {
	for _, length := range []int64{4, -1} {
		sink := &memorySink{}
		handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			one := make([]byte, 1)
			r.Body.Read(one)
			io.WriteString(w, "ok")
		}))
		r := request("1234")
		r.ContentLength = length
		handler.ServeHTTP(httptest.NewRecorder(), r)
		body := find(t, sink, "backend.req.start")
		if body["requestBodyState"] != "truncated" || body["body"] != nil {
			t.Fatal(body)
		}
	}
}
func TestDuplicateCorrelationRejected(t *testing.T) {
	for _, header := range []string{"x-crumbtrail-session-id", "x-crumbtrail-request-id"} {
		sink := &memorySink{}
		r := request("")
		r.Header.Add(header, "second-valid-identity")
		Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(204) })).ServeHTTP(httptest.NewRecorder(), r)
		if len(sink.events()) != 0 {
			t.Fatal("ambiguous request captured")
		}
	}
}
func TestOnlyMatchedRouteTemplateCaptured(t *testing.T) {
	sink := &memorySink{}
	mux := http.NewServeMux()
	mux.HandleFunc("POST /api/users/{email}", func(w http.ResponseWriter, r *http.Request) { io.WriteString(w, "ok") })
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(mux)
	r := request("")
	r.URL.Path = "/api/users/private@example.com"
	handler.ServeHTTP(httptest.NewRecorder(), r)
	encoded, _ := json.Marshal(sink.events())
	if strings.Contains(string(encoded), "private@example.com") {
		t.Fatal("raw path leaked")
	}
	if find(t, sink, "backend.req.start")["route"] != "/api/users/{email}" {
		t.Fatal("matched template missing")
	}
	sink = &memorySink{}
	r = request("")
	r.URL.Path = "/token/secret-path-value"
	Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(http.ResponseWriter, *http.Request) {})).ServeHTTP(httptest.NewRecorder(), r)
	if find(t, sink, "backend.req.start")["url"] != "/" {
		t.Fatal("unmatched raw path recorded")
	}
}

func TestShortDeclaredResponseWithheld(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Content-Length", "4")
		io.WriteString(w, "1")
	}))
	recorder := httptest.NewRecorder()
	handler.ServeHTTP(recorder, request(""))
	if recorder.Body.String() != "1" {
		t.Fatal("response changed")
	}
	end := find(t, sink, "backend.req.end")
	if end["responseBodyState"] != "truncated" || end["responseBody"] != nil {
		t.Fatal(end)
	}
}
func TestHeadDeclaredLengthDoesNotInventTruncation(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Content-Length", "4")
	}))
	r := request("")
	r.Method = "HEAD"
	handler.ServeHTTP(httptest.NewRecorder(), r)
	end := find(t, sink, "backend.req.end")
	if end["responseBodyState"] != "missing" || end["responseBodyTruncated"] != false {
		t.Fatal(end)
	}
}

func TestRequestEOFDoesNotOverrideDeclaredLength(t *testing.T) {
	sink := &memorySink{}
	handler := Middleware(Options{sink, "test", func(*http.Request) bool { return true }})(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { io.ReadAll(r.Body) }))
	r := request("1")
	r.ContentLength = 4
	handler.ServeHTTP(httptest.NewRecorder(), r)
	start := find(t, sink, "backend.req.start")
	if start["requestBodyState"] != "truncated" || start["body"] != nil {
		t.Fatal(start)
	}
}

File: packages/ruby/test/capture_test.rb
require 'minitest/autorun'
require 'rack/mock'
require 'active_record'
require 'crumbtrail/active_record'
require 'stringio'

class CaptureTest < Minitest::Test
  class Sink
    attr_reader :batches
    def initialize; @batches = []; end
    def enqueue(batch); @batches << batch; end
    def events; @batches.flat_map { |batch| batch[:events] }; end
  end
  def setup
    @sink = Sink.new
  end
  def middleware(app, **options)
    Crumbtrail::Middleware.new(app, sink: @sink, service: 'ruby-test', routes: ->(_) { true }, **options)
  end
  def env(body = '{"amount":18.75,"password":"never-send"}')
    Rack::MockRequest.env_for('/api/quote?secret=never-send', method: 'POST', input: body, 'CONTENT_TYPE' => 'application/json').merge(
      'HTTP_X_CRUMBTRAIL_SESSION_ID' => 'ruby-contract-session', 'HTTP_X_CRUMBTRAIL_REQUEST_ID' => 'ruby-contract-request')
  end
  def consume(response)
    status, headers, body = response
    bytes = ''.b
    body.each { |chunk| bytes << chunk }
    body.close if body.respond_to?(:close)
    [status, headers, bytes]
  end
  def test_real_rack_and_database_capture
    ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
    Crumbtrail::ActiveRecord.install(engine: 'sqlite')
    input = nil
    app = middleware(lambda do |e|
      input = e['rack.input'].read
      ActiveRecord::Base.connection.execute('SELECT 731 AS operand')
      begin
        ActiveRecord::Base.connection.execute('SELECT * FROM absent_table')
      rescue ActiveRecord::StatementInvalid
        nil
      end
      [200, { 'content-type' => 'application/json' }, ['{"total":37.5,"currency":"CAD","password":"never-send"}']]
    end)
    response = Rack::MockRequest.new(app).post('/api/quote', input: '{"amount":18.75,"password":"never-send"}', 'CONTENT_TYPE' => 'application/json',
      'HTTP_X_CRUMBTRAIL_SESSION_ID' => 'ruby-contract-session', 'HTTP_X_CRUMBTRAIL_REQUEST_ID' => 'ruby-contract-request')
    assert_equal 200, response.status
    assert_includes input, 'never-send'
    assert_includes response.body, 'never-send'
    assert_nil Crumbtrail.current
    start = @sink.events.find { |e| e[:k] == 'backend.req.start' }[:d]
    ending = @sink.events.find { |e| e[:k] == 'backend.req.end' }[:d]
    assert_equal 18.75, JSON.parse(start[:body])['amount']
    assert_equal 37.5, JSON.parse(ending[:responseBody])['total']
    assert_equal 'redacted', start[:requestBodyState]
    assert @sink.events.any? { |e| e[:k] == 'db.statement' && e[:d][:op] == 'select' }
    assert @sink.events.any? { |e| e[:k] == 'db.error' }
    refute_includes JSON.generate(@sink.batches), 'never-send'
    refute_includes JSON.generate(@sink.batches), 'absent_table'
    File.write(ENV['CRUMBTRAIL_CAPTURE_CONTRACT_OUTPUT'], JSON.generate(@sink.batches)) if ENV['CRUMBTRAIL_CAPTURE_CONTRACT_OUTPUT']
  ensure
    Crumbtrail::ActiveRecord.uninstall
    ActiveRecord::Base.connection_pool.disconnect!
  end
  def test_no_routes_or_bad_correlation_bypass
    app = ->(_) { [204, {}, []] }
    consume(middleware(app, routes: nil).call(env))
    consume(middleware(app).call(env.merge('HTTP_X_CRUMBTRAIL_REQUEST_ID' => "bad\n")))
    assert_empty @sink.events
  end
  def test_truncation_preserves_full_bytes_and_input_rewind
    large = '{"data":"' + 'a' * 20_000 + '"}'
    app = middleware(lambda do |e|
      assert_equal large, e['rack.input'].read
      e['rack.input'].rewind
      assert_equal large, e['rack.input'].read
      [200, { 'content-type' => 'application/problem+json' }, [large]]
    end)
    assert_equal large, consume(app.call(env(large)))[2]
    %w[backend.req.start backend.req.end].each do |kind|
      data = @sink.events.find { |e| e[:k] == kind }[:d]
      assert_equal 'truncated', data[kind.end_with?('start') ? :requestBodyState : :responseBodyState]
    end
  end
  def test_failure_rethrows_and_context_restores
    error = RuntimeError.new('secret-message')
    thrown = assert_raises(RuntimeError) { middleware(->(_) { raise error }).call(env) }
    assert_same error, thrown
    assert_nil Crumbtrail.current
    assert_equal 1, @sink.events.count { |e| e[:k] == 'backend.req.error' }
    refute_includes JSON.generate(@sink.batches), 'secret-message'
  end
  def test_stream_error_and_close_emit_once
    body = Object.new
    def body.each; yield '{'; raise IOError, 'private'; end
    def body.close; end
    response = middleware(->(_) { [200, { 'content-type' => 'application/json' }, body] }).call(env)
    assert_raises(IOError) { consume(response) }
    response[2].close
    assert_equal 1, @sink.events.count { |e| e[:k] == 'backend.req.end' }
    assert_equal 'invalid', @sink.events.last[:d][:responseBodyState]
  end
  def test_broken_sink_preserves_response
    def @sink.enqueue(_); raise 'broken sink'; end
    assert_equal 'ok', consume(middleware(->(_) { [200, {}, ['ok']] }).call(env))[2]
  end
  def test_body_policy
    assert_equal 'invalid', Crumbtrail::Body.capture('{"a":1,"a":2}')[1]
    assert_equal 'invalid', Crumbtrail::Body.capture('{')[1]
    assert_equal 'missing', Crumbtrail::Body.capture('')[1]
    body, state = Crumbtrail::Body.capture('{"amount":18.75,"email":"x@example.com","routing_number":123,"note":"hello world","id":4111111111111111}')
    assert_equal 'redacted', state
    assert_equal 18.75, JSON.parse(body)['amount']
    assert_equal '[REDACTED]', JSON.parse(body)['routing_number']
    refute_includes body, '4111111111111111'
  end
  def test_sender_rejects_unsafe_endpoints
    %w[http://localhost https://user:pass@example.com https://example.com/?key=x].each do |endpoint|
      assert_raises(ArgumentError) { Crumbtrail::Sender.new(endpoint: endpoint, key: 'key') }
    end
  end
end
